packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

```diff
@@ -0,0 +1,234 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { FieldKey, EmptyKey, TreeValue } from "../../tree";
+import {
+	IEditableForest, TreeNavigationResult, mapCursorField, ITreeSubscriptionCursor, ITreeCursor,
+} from "../../forest";
+import { brand } from "../../util";
+import { TreeSchemaIdentifier } from "../../schema-stored";
+
+export const proxySymbol: unique symbol = Symbol("editable-tree proxy");
+
+/**
+ * A unique Symbol to get a method of {@link EditableTree} which returns a type of a node or its children.
+ *
+ * One can call `foo[getTypeSymbol]("bar")` to get a type of `bar` or
+ * without a field name i.e. `foo.bar[getTypeSymbol]()` for the same if `bar` is a non-primitive.
+ */
+export const getTypeSymbol: unique symbol = Symbol("editable-tree getType");
+
+/**
+ * This is a basis type for {@link EditableTree}.
+ *
+ * New features like e.g. to get an iterator over node's fields should be defined here
+ * and for each and every new method its own symbol should be created correspondingly.
+ * This prevents accidental mixin with user-defined string keys.
+ */
+export interface IEditableTreeSignature<T> {
+	[key: string]: T extends number | string | boolean ? T : EditableTreeNode<T>;
+	[getTypeSymbol]: (key?: FieldKey) => TreeSchemaIdentifier;
+}
+
+/**
+ * This converts a type T into an {@link EditableTree} node or a primitive.
+ */
+export type EditableTreeNode<T> = {
+	[P in keyof T]?: T[P] extends number | string | boolean ? T[P] : EditableTree<T[P]>;
+};
+
+/**
+ * A tree which can be traversed and edited.
+ * TODO: support editing.
+ */
+export type EditableTree<T> = IEditableTreeSignature<T> & EditableTreeNode<T>;
```

Comment:
I'm thinking an API like this would work well:

The two symbols might be useful elsewhere as well (If something e else needs them we could move them to the tree folder, though we can leave them here for now).

I've made it non generic: trying to have it do a more specific typed API is really a separate issue (I've been calling that "schema aware APIs", and that will be a separate layer on top of this).

I have also made everything explicitly read-only, since modifying these objects directly is not currently supported as a way to edit.

```suggestion

/**
 * A symbol to for the type of a node in contexts where string keys are already in use for fields.
 */
export const type: unique symbol = Symbol("editable-tree type");

 /**
 * A symbol to for the value of a node in contexts where string keys are already in use for fields.
 */
export const value: unique symbol = Symbol("editable-tree value");

/**
 * A tree which can be traversed and edited.
 *
 * When iterating, only visits non tempt fields.
 * To discover empty fields, inspect the schema using {@link "type"}.
 *
 * TODO: support editing.
 */
export interface EditableTree extends Iterable<EditableField> {
    /**
     * The type of this node.
     * If this node is is well-formed, it must follow this schema.
     */
    readonly [type]: TreeSchema;

    /**
     * Value stored on this node.
     */
    readonly [value]: Value;

    /**
     * Fields of this node, indexed by their field keys (as strings).
     *
     * This API exposes content in a way depending on the {@link Multiplicity} of the {@link FieldKind}.
     * Sequences (including empty ones) are always exposed as arrays,
     * and everything else is either a single EditableTree or undefined depending on if it's empty.
     *
     * TODO:
     * This approach to field lookup can result in collisions between global and local keys,
     * particularly with "extra" fields.
     * A mechanism for disambiguating this should be added,
     * likely involving an alternative mechanism for looking up global fields via symbols.
     */
    readonly [key: string]: readonly EditableTree[] | EditableTree | undefined;
}

/**
 * A field of an {@link EditableTree}.
 */
export type EditableField = readonly [FieldSchema, readonly EditableTree[]];

```

Reply:
this proposed API leaves all the primitive values boxed in their nodes, forcing use of [value] to access them.

If you want to automatically unwrap specific node types directly to their values, you could replace the use of EditiableTree for child access with:

```typescript
/**
 * EditableTree,
 * but with {@link jsonNumber}, {@link jsonBoolean} and {@link jsonString} unwrapped into just their values.
 *
 * TODO:
 * This assumes the schema being used for these trees (which could be a stored schema)
 * matches those specific Json view schema.
 * What happens if they mismatch should be defined.
 *
 * TODO:
 * EditableTree should provide easy access to children in a way thats guaranteed
 * not to do this unwrapping for cases which need to refer to the actual nodes.
 * This may include cases like creating anchors and/or editing.
 */
export type EditableTreeOrPrimitive = EditableTree | number | string | boolean;
```

Reply:
Thanks for your feedback. We are a bit confused now as the API proposed by you is a totally different thing.

Our initial goal was to have an API which mimics a JS object behaviour. If, while traversing the tree, we meet a node having children, we proxify it and dig deeper. If it is a "leaf", it should work just like a generic JS object:
```ts
assert.equal(proxy.name, "Adam");
```
If one would like to get all children (or fields) of the current node, `Object.keys(proxy)` is a way to go.
I have added another test `"traverse a complete tree"` which more or less mocks a behaviour we'd like to get for Inspector.
To be more specific, this type definition allows to have such syntax in tests like `proxy.address.street` etc. without having typecasts all the time one needs to go down to a deeper child:
```
export type EditableTreeNode<T> = {
	[P in keyof T]?: T[P] extends number | string | boolean ? T[P] : EditableTree<T[P]>;
};
```
whereas the second part (`IEditableTreeSignature`) safeguards our proxy against not allowed fields and key collisions.

We could have the proposed API implemented as well, but that wouldn't fit our needs.

Reply:
> I have also made everything explicitly read-only, since modifying these objects directly is not currently supported as a way to edit.

That's safeguarded with an assertions and tests, also we'd like to start implementing edit like yesterday =)

Overall I have an impression that it is forgotten that this PR is a WiP (in general) and 1-2 other PRs (like UI support, schematised API and edit support) will immediately follow, meanwhile improving this by using it in a "real" implementation of other features. We hoped that this would be an iterative process.